### PR TITLE
[CBRD-26404] Add test case for IS NOT NULL incorrectly removed in OUTER JOINs for NOT NULL columns from the inner table

### PR DESCRIPTION
**Pull request by kimhak-tech**
Refer to http://jira.cubrid.org/browse/CBRD-26404

**CBRD-26404** fixed an issue where IS NOT NULL conditions on NOT NULL columns are incorrectly removed in OUTER JOIN queries.

### CODE REVIEW

File: sql/_13_issues/_26_1h/cases/cbrd_26404.sql [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+INSERT INTO t3 VALUES (1,1), (2,2), (5,5);
+
+evaluate 'Case 5: Multiple OUTER JOINs - IS NOT NULL on inner tables';
+SELECT /*+ recompile */ a.*, b.*, c.*
+FROM t1 a
+LEFT OUTER JOIN t2 b ON a.c1 = b.c1
+LEFT OUTER JOIN t3 c ON a.c1 = c.c1
+WHERE b.c2 IS NOT NULL
+  AND c.c2 IS NOT NULL;
+
```
Comment (kwonhoil) [inline sql/_13_issues/_26_1h/cases/cbrd_26404.sql:88]: # case 1
select * from
(
select c1, c2 from t1  union all
select c1, c2 from t2  union all
select c1, c2 from t3 union all
select null, null from db_root 
)
where c2 is not null;

# case 2
select * from
(
select c1, c2 from t1  union all
select c1, c2 from t2  union all
select c1, c2 from t3 
)
where c2 is not null;

# case 3
select * from
(
select c1, c2 from t1  union all
select c1, c2 from t2 where c2 is not null union all
select c1, c2 from t3 
)
;

File: sql/_13_issues/_26_1h/cases/cbrd_26404.sql [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+DROP TABLE IF EXISTS t1;
+DROP TABLE IF EXISTS t2;
+
+CREATE TABLE t1(c1 INT, c2 INT NOT NULL);
+INSERT INTO t1 VALUES(1, 1), (2, 2), (3, 3);
+
+CREATE TABLE t2(c1 INT, c2 INT NOT NULL);
+INSERT INTO t2 VALUES(2, 2), (4, 4);
+
+evaluate 'Case 1: LEFT OUTER JOIN - IS NOT NULL on inner table (t2.c2)';
```
Comment (ssihil) [inline sql/_13_issues/_26_1h/cases/cbrd_26404.sql:16]: Add oracle style outer join queries for each scenario.

SELECT /*+ recompile */ a.\*, b.\*
FROM t1 a, t2 b WHERE a.c1 = b.c1(+)
and b.c2 IS NOT NULL;

File: sql/_13_issues/_26_1h/cases/cbrd_26404.sql (hunk 1 above)
Comment (kimhak-tech) [reply sql/_13_issues/_26_1h/cases/cbrd_26404.sql:88]: All above three cases have included in the test case.

File: sql/_13_issues/_26_1h/cases/cbrd_26404.sql (hunk 2 above)
Comment (kimhak-tech) [reply sql/_13_issues/_26_1h/cases/cbrd_26404.sql:16]: Reflected!

**Review by ssihil — APPROVED** [review]

File: sql/_13_issues/_26_1h/cases/cbrd_26404.sql [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+) WHERE c2 IS NOT NULL;
+
+evaluate 'Case 8: UNION ALL with IS NOT NULL in one branch';
+SELECT * FROM (
+  SELECT c1, c2 FROM t1
+  UNION ALL
+  SELECT c1, c2 FROM t2 WHERE c2 IS NOT NULL
+  UNION ALL
+  SELECT c1, c2 FROM t3
+);
```
Comment (kwonhoil) [inline sql/_13_issues/_26_1h/cases/cbrd_26404.sql]: SELECT * FROM (
  SELECT 't1' c0, c1, c2 FROM t1
  UNION ALL
  SELECT 't2' c0, c1, c2 FROM t2
  UNION ALL
  SELECT 't3' c0, c1, c2 FROM t3
  UNION ALL
  SELECT 't4' c0, NULL, NULL FROM db_root
) WHERE c2 IS NOT NULL 
  ORDER BY c0, c1;

evaluate 'Case 7: UNION ALL without null row - IS NOT NULL filter';
SELECT * FROM (
  SELECT 't1' c0, c1, c2 FROM t1
  UNION ALL
  SELECT 't2' c0, c1, c2 FROM t2
  UNION ALL
  SELECT 't3' c0, c1, c2 FROM t3
) WHERE c2 IS NOT NULL
  ORDER BY c0, c1;

evaluate 'Case 8: UNION ALL with IS NOT NULL in one branch';
SELECT * FROM (
  SELECT 't1' c0, c1, c2 FROM t1
  UNION ALL
  SELECT 't2' c0, c1, c2 FROM t2 WHERE c2 IS NOT NULL
  UNION ALL
  SELECT 't3' c0, c1, c2 FROM t3
)
  ORDER BY c0, c1;

Comment (kimhak-tech) [reply sql/_13_issues/_26_1h/cases/cbrd_26404.sql]: Reflected!

**Review by kwonhoil — APPROVED** [review]

**Review by junsklee — APPROVED** [review]

File: sql/_13_issues/_26_1h/cases/cbrd_26404.sql [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+  AND a.c2 IS NOT NULL
+  AND b.c2 IS NOT NULL;
+
+evaluate 'Case 4: IS NOT NULL with additional condition';
+SELECT /*+ recompile */ a.*, b.*
+FROM t1 a
+LEFT OUTER JOIN t2 b ON a.c1 = b.c1
+WHERE b.c2 IS NOT NULL
+  AND b.c1 > 1
+ORDER BY a.c1;
```
Comment (bagus-kim) [inline sql/_13_issues/_26_1h/cases/cbrd_26404.sql:66]: Case4만 order by가 있네요 다른 case에도 추가하는 것이 좋지 않나요?

File: sql/_13_issues/_26_1h/cases/cbrd_26404.sql [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+evaluate 'Case 8: UNION ALL with IS NOT NULL in one branch';
+SELECT * FROM (
+  SELECT 't1' c0, c1, c2 FROM t1
+  UNION ALL
+  SELECT 't2' c0, c1, c2 FROM t2 WHERE c2 IS NOT NULL
+  UNION ALL
+  SELECT 't3' c0, c1, c2 FROM t3
+)
+ORDER BY c0, c1;
+
```
Comment (bagus-kim) [inline sql/_13_issues/_26_1h/cases/cbrd_26404.sql:129]: 하단의 case추가를 고려해 주세요

--evaluate 'Add1: LEFT OUTER JOIN - IS NOT NULL in ON clause'; on절에 is not null인 경우
SELECT /*+ recompile */ a.*, b.*
FROM t1 a
LEFT OUTER JOIN t2 b
  ON a.c1 = b.c1 AND b.c2 IS NOT NULL
ORDER BY a.c1;


--evaluate 'Add2: Derived table inner - IS NOT NULL must not be removed'; subquery사용 시
SELECT /*+ recompile */ a.*, b.*
FROM t1 a
LEFT OUTER JOIN (SELECT * FROM t2) b ON a.c1 = b.c1
WHERE b.c2 IS NOT NULL
ORDER BY a.c1;

--evaluate 'Add3: LEFT OUTER JOIN with index - IS NOT NULL must remain correct'; 인덱스가 존재하는 경우 
CREATE INDEX idx_t2_c1 ON t2(c1);

SELECT /*+ recompile */ a.*, b.*
FROM t1 a
LEFT OUTER JOIN t2 b ON a.c1 = b.c1
WHERE b.c2 IS NOT NULL
ORDER BY a.c1;

**Review by tw-kang — APPROVED** [review]

File: sql/_13_issues/_26_1h/cases/cbrd_26404.sql (hunk 4 above)
Comment (kimhak-tech) [reply sql/_13_issues/_26_1h/cases/cbrd_26404.sql:66]: Reflected.

File: sql/_13_issues/_26_1h/cases/cbrd_26404.sql (hunk 5 above)
Comment (kimhak-tech) [reply sql/_13_issues/_26_1h/cases/cbrd_26404.sql:129]: Reflected.

**Review by bagus-kim — APPROVED** [review]

**Review by Hamkua — APPROVED** [review]
